src/Storages/ObjectStorage/StorageObjectStorage.h

```diff
@@ -0,0 +1,193 @@
+        const StorageMetadataPtr & metadata_snapshot,
+        ContextPtr local_context,
+        TableExclusiveLockHolder &) override;
+
+    bool supportsPartitionBy() const override { return true; }
+
+    bool supportsSubcolumns() const override { return true; }
+
+    bool supportsTrivialCountOptimization(const StorageSnapshotPtr &, ContextPtr) const override { return true; }
+
+    bool supportsSubsetOfColumns(const ContextPtr & context) const;
+
+    bool prefersLargeBlocks() const override;
+
+    bool parallelizeOutputAfterReading(ContextPtr context) const override;
+
+    static SchemaCache & getSchemaCache(const ContextPtr & context, const std::string & storage_type_name);
+
+    static ColumnsDescription resolveSchemaFromData(
+        const ObjectStoragePtr & object_storage,
+        const ConfigurationPtr & configuration,
+        const std::optional<FormatSettings> & format_settings,
+        const ContextPtr & context);
+
+    static std::string resolveFormatFromData(
+        const ObjectStoragePtr & object_storage,
+        const ConfigurationPtr & configuration,
+        const std::optional<FormatSettings> & format_settings,
+        const ContextPtr & context);
+
+    static std::pair<ColumnsDescription, std::string> resolveSchemaAndFormatFromData(
+        const ObjectStoragePtr & object_storage,
+        const ConfigurationPtr & configuration,
+        const std::optional<FormatSettings> & format_settings,
+        const ContextPtr & context);
+
+protected:
+    virtual void updateConfiguration(ContextPtr local_context);
+
+    static std::unique_ptr<ReadBufferIterator> createReadBufferIterator(
+        const ObjectStoragePtr & object_storage,
+        const ConfigurationPtr & configuration,
+        const std::optional<FormatSettings> & format_settings,
+        ObjectInfos & read_keys,
+        const ContextPtr & context);
+
+    ConfigurationPtr configuration;
+    const ObjectStoragePtr object_storage;
+    const std::string engine_name;
+    const std::optional<FormatSettings> format_settings;
+    const ASTPtr partition_by;
+    const bool distributed_processing;
+
+    LoggerPtr log;
+};
+
+class StorageObjectStorage::Configuration
+{
+public:
+    Configuration() = default;
+    Configuration(const Configuration & other);
+    virtual ~Configuration() = default;
+
+    using Path = std::string;
+    using Paths = std::vector<Path>;
+
+    static void initialize(
+        Configuration & configuration,
+        ASTs & engine_args,
+        ContextPtr local_context,
+        bool with_table_structure);
+
+    virtual std::string getTypeName() const = 0;
+    virtual std::string getEngineName() const = 0;
+
+    virtual Path getPath() const = 0;
+    virtual void setPath(const Path & path) = 0;
+
+    virtual const Paths & getPaths() const = 0;
+    virtual void setPaths(const Paths & paths) = 0;
+
+    virtual String getDataSourceDescription() = 0;
+    virtual String getNamespace() const = 0;
```

Comment:
What is namespace? Would be good to have a comment

Reply:
A bucket, for example. Yes will add a comment. BTW now sure if namespace is a good naming for it, but I named it this way for consistency with IObjectStorage: https://github.com/ClickHouse/ClickHouse/blob/3af24dacf89250ef130024a69954d4c7fafc36f2/src/Disks/ObjectStorages/IObjectStorage.h#L199-L201